Make MinimizerResult API more similar to ModelResult

tritemio opened the pull request:
In the process of adding support for MinimizerResult in [pybroom](https://github.com/tritemio/pybroom) I found some little inconsistencies between MinimizerResult and ModelResult in lmfit.

This PR makes the interface of the two classes more similar by adding to MinimizerResults two attributes: the dict `init_values` and string `method`.

newville commented:
@tritemio Thanks -- I agree with much of this:

* support 'least_squares'
* add `method` name and `init_values` to `MinimizerResult`

but: 
 * dict comprehension is not supported in Py 2.6 and so cannot be used until lmfit 0.9.5.
 * add `least_squares` to docstring for `minimize()` function.
 * I'd rather not change the call signature to `prepare_fit()` (especially to change `params` from optional keyword to required positional argument).  If `method` is added to the call signature, it should be a keyword argument.  But, I'd rather *not* set the method name in `prepare_fit`, as it sort of looks like one would expect that to dictate which method was going to be used, and that is not the case.   So, better (for now, at least) to set after the `MinimizerResult` is created, and possibly even after the fit is completed.

andyfaff commented on doc/fitting.rst:
```diff
@@ -15,59 +15,17 @@ when using complicated constraints or comparing results from related fits.
-===============================
+=============================
+
+.. currentmodule:: lmfit.minimizer
```
please leave as:

    ..currentmodule:: minimizer

This makes the intersphinx references uniform across parameters.rst and model.rst. For example, the intersphinx reference for `Parameters` is currently `lmfit::Parameters` and `minimize` is `lmfit::minimize`.
If this change is kept then the intersphinx reference becomes `lmfit::lmfit.minimize`.

andyfaff replied:
If we do make the change, then all the module references in the other documentation files should be changed to match this.

tritemio replied:
With 

```
..currentmodule:: lmfit.minimizer
```

sphinx creates a link to a function using the intuitive syntax:

```
:func:`minimize`
```

When I change the directive to 

```
..currentmodule:: minimizer
```

the link is not created anymore. Is there a different syntax in this case?

In general, I would favor cleaner/shorter internal links over shorter intersphinx links.

andyfaff replied:
> In general, I would favor cleaner/shorter internal links over shorter
intersphinx links

OK.  One of the problems with lmfit documentation in general is that the
external intersphinx links aren't consistent. E.g. one might have to use
`lmfit::Parameters` but `lmfit.Minimizer.minimize`. It's impossible to
predict which one. One has to inspect the objects.inv file to see which one
to use.

tritemio replied:
agree, see my comment in https://github.com/lmfit/lmfit-py/issues/354

andyfaff commented on lmfit/minimizer.py:
```diff
@@ -1369,40 +1377,65 @@ def _nan_policy(a, nan_policy='raise', handle_inf=True):
+    to be minimized, a dictionary (:class:`lmfit.parameter.Parameters`)
+    containing the model parameters, and several optional arguments.
+
+    Notes
```
The notes section should follow the `Parameters` and `Returns` sections, it shouldn't be at the top.

tritemio replied:
ok

andyfaff commented on lmfit/minimizer.py:
```diff
@@ -1369,40 +1377,65 @@ def _nan_policy(a, nan_policy='raise', handle_inf=True):
-        `fcn(params, *args, **kws)`
-    params : lmfit.parameter.Parameters object.
+        objective function to be minimized. When method is `leastsq` or
+        `least_squares`, the objective function need to return an array
```
needs

tritemio replied:
thanks

andyfaff commented on lmfit/minimizer.py:
```diff
@@ -1369,40 +1377,65 @@ def _nan_policy(a, nan_policy='raise', handle_inf=True):
+        objective function to be minimized. When method is `leastsq` or
+        `least_squares`, the objective function need to return an array
+        of residuals (difference between model and data) to be minimized
+        in a least squares sense.  With other scalar methods the objective
```
`Minimizer.penalty` allows a scalar value *or* the array of residuals to be returned. It doesn't need to be a scalar.

    `With the scalar methods the objective function can either return the residuals array or a single scalar value`

tritemio replied:
done

andyfaff commented on lmfit/minimizer.py:
```diff
@@ -1415,20 +1448,28 @@ def minimize(fcn, params, method='leastsq', args=None, kws=None,
     fit_kws : dict, optional
         Options to pass to the minimizer being used.
 
-    Notes
```
Why were the notes placed at the top of the docstring? When using interactive help people want to see Parameters first, then Returns, then Notes.

tritemio replied:
ok

andyfaff commented on lmfit/minimizer.py:
```diff
@@ -1415,20 +1448,28 @@ def minimize(fcn, params, method='leastsq', args=None, kws=None,
+        	           iter_cb=iter_cb, scale_covar=scale_covar, **fit_kws)
+        fitter.minimize(method=method)
 
-    A common use for `args` and `kwds` would be to pass in other
```
Why are these lines removed?

tritemio replied:
They are in the notes section

tritemio commented:
I fixed the 3 issues raised by @newville 

Moreover, I added all the changes suggested by @andyfaff except for the currentmodule directive which breaks internal links (see my inline comment).

Since this PR also modifies `Minimizer.minimize` and `MinimizerResult`, I also expanded  the use autodoc to these two items (a class and a method) in the process of documenting new attributes and methods.

Other than that, I think we should address the decision on the currentmodule syntax and the completion of autodoc use (the entire Minimizer class should be autodoc-ed) to another PR and preferably continue the discussion in #354.

So, if there are no more outstanding issues related to the topic of this PR I would suggest merging it and moving further general docs discussion to #354 to be addressed in a new PR.

tritemio commented:
@newville, I'm surprised and disappointed you marked the 0.9.4 release without including this PR. Any reasons? I need this PR to simplify the job of pybroom and I wanted to support 0.9.4+, but I can't now.

newville commented:
@tritemio Oh, very sorry. My mistake -- somehow in the discussion of sphinx markup I completely missed that you had fixed this.    So the only reason is my haste and carelessness.

That said, would merging this and releasing 0.9.5 now be sufficient or did you need this and pybroom to work with Python2.6?   I have no objection to doing that.   There are other cleanups of code and docs (#332 and #354)  I'd like to see, but I won't have time to work on these for a while.

newville commented:
@tritemio  merged this branch (by hand) to master.

tritemio commented:
@newville, no problem, mistakes happens.

Yes a 0.9.5 release would fix any issue for me, I don't plan to support py2 in pybroom anyway.

Let me check what you merged, I may have some more docs fixes.

newville commented:
@tritemio  Thanks for understanding, and apologies again.  I'll close this one, and we can continue from here.   Shall we aim for 0.9.5 release with your fixes on Tuesday?
